### PR TITLE
fix(kademlia): increase bit suffix length to improve bin population

### PR DESCRIPTION
**Pull request by istae**
### Checklist

- [ ] I have read the [coding guide](https://github.com/ethersphere/bee/blob/master/CODING.md)
- [ ] My change requires a documentation update and I have done it
- [ ] I have added tests to cover my changes.

### Screenshots (if appropriate):

---
This change is [Reviewable](https://reviewable.io/reviews/ethersphere/bee/2971)

### CODE REVIEW

File: pkg/topology/kademlia/kademlia.go [hunk 1]
```diff
@@ -281,18 +281,15 @@ func (k *Kad) connectBalanced(wg *sync.WaitGroup, peerConnChan chan<- *peerConnI
 	}
 }
 
-// connectNeighbours attempts to connect to the neighbours
-// which were not considered by the connectBalanced method.
-func (k *Kad) connectNeighbours(wg *sync.WaitGroup, peerConnChan chan<- *peerConnInfo) {
+// connectUntilSaturation attempts to connect to peers until bin oversaturation is reached.
+func (k *Kad) connectUntilSaturation(wg *sync.WaitGroup, peerConnChan chan<- *peerConnInfo) {
```
Comment (mrekucci) [inline pkg/topology/kademlia/kademlia.go:285]: I recommend calling it `connectUntilOversaturation`.

**Comment by acud** [comment]
don't understand.... didn't we say we're just changing the saturation and oversaturation parameter values? btw inside the neighborhood you will always try to connect to all neighbors, so concepts of saturation and oversaturation are largely irrelevant. saturation is only relevant **outside** of neighborhood

**Comment by istae** [comment]
@acud please ignore the draft PR for now 👍

**Comment by istae** [comment]
> don't understand.... didn't we say we're just changing the saturation and oversaturation parameter values? btw inside the neighborhood you will always try to connect to all neighbors, so concepts of saturation and oversaturation are largely irrelevant. saturation is only relevant **outside** of neighborhood

having looked at kademlia, we actually don't use the `saturationPeers` anywhere meaningfully. This PR changes that and also increases the saturation limits.

File: pkg/topology/kademlia/kademlia.go [hunk 2]
```diff
@@ -281,23 +281,22 @@ func (k *Kad) connectBalanced(wg *sync.WaitGroup, peerConnChan chan<- *peerConnI
...
 	var currentPo uint8 = 0
 
-	_ = k.knownPeers.EachBinRev(func(addr swarm.Address, po uint8) (bool, bool, error) {
-		depth := k.NeighborhoodDepth()
+	_ = k.knownPeers.EachBinRev(func(addr swarm.Address, bin uint8) (bool, bool, error) {
 
-		// out of depth, skip bin
-		if po < depth {
+		// if the bin is already saturated AND the bin is out of the depth, skip bin
+		if bin < k.NeighborhoodDepth() && k.connectedPeers.BinSize(bin) >= saturationPeers {
```
Comment (acud) [inline pkg/topology/kademlia/kademlia.go:294]: again, there is no notion of saturation within depth and we always want to connect to all peers

**Review by acud — CHANGES_REQUESTED** [review]

File: pkg/topology/kademlia/kademlia.go (hunk 1 above)
Comment (acud) [reply pkg/topology/kademlia/kademlia.go:286]: I find this change unnecessary. We're mixing things up here. The parameter change should be enough, please confirm @istae

Comment (acud) [reply pkg/topology/kademlia/kademlia.go:286]: If the connector method handles both bins which are inside and outside of depth there can not exist any notion of saturation in the name

Comment (istae) [reply pkg/topology/kademlia/kademlia.go:286]: So how's `connectToPeers` ?

Comment (acud) [reply pkg/topology/kademlia/kademlia.go:286]: this is unfortunately not a naming issue, but an issue of single responsibility. you're not just renaming but mixing up responsibilities. the change you've done here basically overrides the responsibility of the bin balancer by connecting to peers outside of depth, which would mess up the balancing inside the bins.

Comment (acud) [reply pkg/topology/kademlia/kademlia.go:286]: > the neighborhood connecter ignores bins that are out of depth

that is incorrect.

have a look at this conditional that you wrote:
```go
		if bin < k.NeighborhoodDepth() && k.connectedPeers.BinSize(bin) >= saturationPeers {
```
this means that if the bin is out of depth and unsaturated then this connector will indeed cause more peers to be connected outside of depth and this is not what we want. we always want the balancer to do connections out of depth. no reinforcements needed there. if you want more peers out of depth you need to change the number of bits the balancer is trying to brute force.

Comment (istae) [reply pkg/topology/kademlia/kademlia.go:286]: Without a new connector function, the issue of some of the shallower bins never filling up won't be resolved.

Comment (istae) [reply pkg/topology/kademlia/kademlia.go:286]: > the neighborhood connecter ignores bins that are out of depth
yes, `connectNeighbours` in master is doing this,.

**Review by mrekucci — APPROVED** [review]

**Review by acud — APPROVED** [review]

**Pull request merged**